### PR TITLE
Mobx: Senaps locations catalog item

### LINKED ISSUE #3897: Model architecture: Implement Senaps catalog item

### PR DESCRIPTION
**Pull request by rowanwins**
### What this PR does
Adds support for displaying Senaps locations and associated streams.

screenshot-localhost_3001-2020 04 23-12_47_38

Fixes #3897

Configure with something like
````
    {
      "name": "Senaps Locations",
      "type": "senaps-locations",
      "locationIdFilter": "*boorowa*",
      "streamIdFilter": "*SHT31DIS_ALL*",
      "style": {
        "marker-color": "red"
      }
    }
````

This also requires a senaps api key configured with terria-server (v3.2.0 or above)
````
    "appendParamToQueryString": {
        "senaps.io": [
            {
                "regexPattern": ".",
                "params": {
                    "apikey": "123"
                }
            }
        ]
    },
````

I've also included a couple of little fixes for our CSV handling
- [x] bumped up our papaparse version because the old one was requesting CSV's with a strange header which was getting an error back from Senaps
- [x]  adding some improved logic to our chunking
  - basically senaps can't supply a Content-Length header which means when we were making our chunking requests with papaparse it wouldn't know when to finish and would throw and error when it made an invalid request because it had gone to high. I've added an extra check to see if we've received some response, and if so, use what we do have, rather than freaking out and showing nothing.

### Checklist
-   [x] There are unit tests to verify my changes are correct
-   [x] I've updated CHANGES.md with what I changed.

### CODE REVIEW

File: lib/Models/SenapsLocationsCatalogItem.ts [hunk 1]
```diff
@@ -0,0 +1,279 @@
...
+
+export interface SenapsFeature {
+  type: string;
+  properties: {
+    id: string;
+    description: string;
+    endpoint: string;
+    hasStreams: boolean | null;
+    streamIds: string[];
+    streamId: string | null;
```
Comment (KeyboardSounds) [inline lib/Models/SenapsLocationsCatalogItem.ts]: Out of curiosity, why is there streamId and streamIds? If there's only one streamId, shouldn't it just be streamIds[0]? Is it a weird Senaps thing?

File: lib/Models/SenapsLocationsCatalogItem.ts [hunk 2]
```diff
@@ -0,0 +1,279 @@
...
+    private readonly senapsLocationsCatalogItem: SenapsLocationsCatalogItem,
+    private readonly geojsonItem: GeoJsonCatalogItem
+  ) {
+    super();
+    this.geojsonItem = geojsonItem;
+  }
+
+  duplicateLoadableStratum(newModel: BaseModel): this {
+    return new SenapsLocationsStratum(
+      this.senapsLocationsCatalogItem,
```
Comment (KeyboardSounds) [inline lib/Models/SenapsLocationsCatalogItem.ts]: Don't you need to pass newModel to the duplicate stratum?

**Review by KeyboardSounds — CHANGES_REQUESTED** [review]
Nice job deciphering those rather cryptic API docs! This is looking really good. Most of my comments are only stylistic things, or asking for clarification of stuff I don't understand.

File: lib/Models/SenapsLocationsCatalogItem.ts [hunk 3]
```diff
@@ -0,0 +1,279 @@
+import i18next from "i18next";
+import { computed, runInAction } from "mobx";
+import URI from "urijs";
+
+import Cartesian3 from "terriajs-cesium/Source/Core/Cartesian3";
+import Cartographic from "terriajs-cesium/Source/Core/Cartographic";
+import Ellipsoid from "terriajs-cesium/Source/Core/Ellipsoid";
+import JulianDate from "terriajs-cesium/Source/Core/JulianDate";
+import Property from "terriajs-cesium/Source/Core/Property";
+import sampleTerrain from "terriajs-cesium/Source/Core/sampleTerrain";
```
[the remark this answers is not in the document]
Comment (rowanwins) [reply lib/Models/SenapsLocationsCatalogItem.ts]: cut-and-pasta :)

File: lib/Models/SenapsLocationsCatalogItem.ts [hunk 4]
```diff
@@ -0,0 +1,279 @@
...
+      title='{{id}}'
+      sources='https://senaps.io/api/sensor/v2/observations?streamid={{#terria.urlEncodeComponent}}{{streamIds}}{{/terria.urlEncodeComponent}}&limit=1440&media=csv&csvheader=false&sort=descending,https://senaps.io/api/sensor/v2/observations?streamid={{#terria.urlEncodeComponent}}{{streamIds}}{{/terria.urlEncodeComponent}}&limit=7200&media=csv&csvheader=false&sort=descending'
+      source-names='1d,5d'
+      downloads='https://senaps.io/api/sensor/v2/observations?streamid={{#terria.urlEncodeComponent}}{{streamIds}}{{/terria.urlEncodeComponent}}&limit=1440&media=csv&csvheader=false&sort=descending,https://senaps.io/api/sensor/v2/observations?streamid={{#terria.urlEncodeComponent}}{{streamIds}}{{/terria.urlEncodeComponent}}&limit=7200&media=csv&csvheader=false&sort=descending'
+      download-names='1d,5d'
+    >
+    </chart>
+  {{/hasStreams}}
+  {{^hasStreams}}
+    No streams ar this locations
```
[the remark this answers is not in the document]
Comment (rowanwins) [reply lib/Models/SenapsLocationsCatalogItem.ts]: 👍

File: lib/Models/SenapsLocationsCatalogItem.ts [hunk 5]
```diff
@@ -0,0 +1,279 @@
...
+              streamId: null
+            },
+            geometry: site.geojson
+          };
+          addStreamIds(f, i);
+          return f;
+        })
+      };
+
+      const gjci = new GeoJsonCatalogItem(
```
[the remark this answers is not in the document]
Comment (rowanwins) [reply lib/Models/SenapsLocationsCatalogItem.ts]: 👍

File: lib/Models/SenapsLocationsCatalogItem.ts [hunk 6]
```diff
@@ -0,0 +1,279 @@
...
+      );
+      gjci.setTrait("definition", "geoJsonData", (fc as any) as JsonObject);
+
+      if (isDefined(senapsLocationsCatalogItem.style)) {
+        gjci.setTrait("definition", "style", senapsLocationsCatalogItem.style);
+      }
+
+      const featureInfo = createStratumInstance(FeatureInfoTemplateTraits, {
+        template: `<h4>Site: {{id}}</h4>
+  <h5 style="margin-bottom:5px;">Available Streams</h5>
```
[the remark this answers is not in the document]
Comment (rowanwins) [reply lib/Models/SenapsLocationsCatalogItem.ts]: Good thinking - I didn't think of that!

File: lib/Models/SenapsLocationsCatalogItem.ts [hunk 7]
```diff
@@ -0,0 +1,279 @@
...
+      });
+    }
+    return [];
+  }
+
+  protected forceLoadMetadata(): Promise<void> {
+    return Promise.resolve();
+  }
+
+  _constructLocationsUrl() {
```
[the remark this answers is not in the document]
Comment (rowanwins) [reply lib/Models/SenapsLocationsCatalogItem.ts:264]: Yeah one of those - i got confused making my tests pass and having them callable from within my test, I'l relook at it

File: lib/Models/SenapsLocationsCatalogItem.ts (hunk 1 above)
Comment (rowanwins) [reply lib/Models/SenapsLocationsCatalogItem.ts]: I use `streamId` for the default chart shown in the popup which isn't ideal, I'll relook at it because perhaps with my refactoring I can show multiple streams on my chart now...

File: lib/Models/SenapsLocationsCatalogItem.ts (hunk 2 above)
Comment (rowanwins) [reply lib/Models/SenapsLocationsCatalogItem.ts]: ¯\_(ツ)_/¯ that's where my knowledge starts running out in mobx world :)

File: lib/Models/SenapsLocationsCatalogItem.ts (hunk 1 above)
Comment (rowanwins) [reply lib/Models/SenapsLocationsCatalogItem.ts]: Have culled streamId - no longer required

File: lib/Models/SenapsLocationsCatalogItem.ts (hunk 7 above)
[the remark this answers is not in the document]
Comment (rowanwins) [reply lib/Models/SenapsLocationsCatalogItem.ts:264]: So if I set to `protected` or `private` I can't access the methods in the  `SenapsLocationsStratum` class where I currently am . 

Because they're currently being used in the `static` `load` method I don't yet access to the full  `SenapsLocationsStratum` as `this` or whatver.

So I'm inclined to leave it as is.

[the remark this answers is not in the document]
Comment (KeyboardSounds) [reply lib/Models/SenapsLocationsCatalogItem.ts:264]: Sounds fine to me!

**Review by KeyboardSounds — APPROVED** [review]
